# Numpy

In [1]:
import numpy as np

In [2]:
L = [1,2,3]
A = np.array([1,2,3])

In [3]:
for e in L:
    print e

1
2
3


In [4]:
for e in A:
    print e

1
2
3


In [5]:
L.append(4)
L

[1, 2, 3, 4]

In [6]:
L = L + [5]
L

[1, 2, 3, 4, 5]

In [7]:
L2 = []
for e in L:
    L2.append(e + e)
L2

[2, 4, 6, 8, 10]

Numpy use vectors, not lists

In [8]:
A + A

array([2, 4, 6])

In [9]:
2 * L

[1, 2, 3, 4, 5, 1, 2, 3, 4, 5]

In [10]:
2 * A

array([2, 4, 6])

## Element wise squaring with python lists and Numpy

In [11]:
L2 = []
for e in L:
    L2.append(e * e)
L2

[1, 4, 9, 16, 25]

In [12]:
A ** 2

array([1, 4, 9])

## Elementwise square root

In [13]:
np.sqrt(A)

array([ 1.        ,  1.41421356,  1.73205081])

## Elementwise log

In [14]:
np.log(A)

array([ 0.        ,  0.69314718,  1.09861229])

## Elementwise exponential

In [15]:
np.exp(A)

array([  2.71828183,   7.3890561 ,  20.08553692])

# Dot product
With the definition of dot product, we multiply the vectors and sum the results

In [16]:
a = np.array([1,2])
b = np.array([2,1])
dot = 0
for e, f in zip(a, b):
    dot += e * f
dot

4

Also, with numpy arrays we can do elementwise multiplication and use the sum method to get the dot product

In [17]:
np.sum(a * b)

4

Or we can use the dot method of numpy

In [18]:
np.dot(a, b)

4

In [19]:
a.dot(b)

4

Let's compare how many times faster the numpy method is compared to the python list method.

In [20]:
from datetime import datetime

a = np.random.randn(100)
b = np.random.randn(100)
T = 100000

def slow_dot_product(a, b):
    result = 0
    for e, f in zip(a, b):
        result += e*f
    return result

t0 = datetime.now()
for t in xrange(T):
    slow_dot_product(a, b)
dt1 = datetime.now() - t0

t0 = datetime.now()
for t in xrange(T):
    a.dot(b)
dt2 = datetime.now() - t0

print "dt1 / dt2: ", dt1.total_seconds() / dt2.total_seconds()

dt1 / dt2:  58.3520260868


# Matrixes
Python lists vs numpy matrixes

In [21]:
M = np.array([[1,2],[3,4]])
L = [[1,2],[3,4]]

In [22]:
M

array([[1, 2],
       [3, 4]])

In [23]:
L

[[1, 2], [3, 4]]

Access a value with lists

In [24]:
L[0][0]

1

Access a value with numpy matrix

In [25]:
M[0,0]

1

# Generating Matrixes to work with

## Array of zeros

In [26]:
Z = np.zeros(10)
Z

array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.])

In [27]:
Z = np.zeros((10,10))
Z

array([[ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.]])

## Array of ones

In [28]:
O = np.ones(10)
O

array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])

In [29]:
O = np.ones((10, 10))
O

array([[ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.]])

## Array of random numbers
This method gives us numbers between 0 and 1

In [30]:
R = np.random.random(10)
R

array([ 0.51360568,  0.44648265,  0.86447103,  0.1030727 ,  0.99218608,
        0.76987159,  0.22708916,  0.90133688,  0.59653665,  0.76388331])

In [31]:
R = np.random.random((10, 10))
R

array([[ 0.32209721,  0.6638846 ,  0.66126171,  0.78638955,  0.36410112,
         0.13004819,  0.72191261,  0.30210821,  0.97784494,  0.78264644],
       [ 0.07427585,  0.76571014,  0.73122005,  0.72078447,  0.26492989,
         0.38254658,  0.98249619,  0.81815785,  0.87689965,  0.21814196],
       [ 0.26246999,  0.56541023,  0.73959354,  0.35563112,  0.81654008,
         0.06032338,  0.19887188,  0.60961057,  0.73176913,  0.42047766],
       [ 0.92163152,  0.26532809,  0.57264164,  0.45409052,  0.08569349,
         0.65610004,  0.98030046,  0.63332203,  0.36878498,  0.65381564],
       [ 0.13396699,  0.2470549 ,  0.67948861,  0.06139543,  0.6404528 ,
         0.64040647,  0.07948241,  0.619944  ,  0.78479739,  0.58802405],
       [ 0.06536396,  0.49002001,  0.30183904,  0.75707349,  0.18979716,
         0.94962499,  0.53347207,  0.11242954,  0.73028323,  0.42968431],
       [ 0.79081441,  0.03366207,  0.1575876 ,  0.75070354,  0.62923384,
         0.23486969,  0.22779682,  0.84746278

Gaussian distributed numbers

In [32]:
G = np.random.randn(10,10)
G

array([[-0.86582089,  1.26163605, -0.52903998, -2.34170292, -1.34282619,
        -1.60014597, -2.04709803, -0.45925924, -0.73215275,  0.04812003],
       [ 0.55395138, -0.41908858,  0.29135729,  1.08825589, -1.12834213,
         0.09665552,  1.59109152, -1.65775696, -0.20033698, -1.64446549],
       [ 0.06062332, -1.14912674, -0.16370373,  0.91138332,  0.19511162,
         1.65882576, -0.07801294, -0.28388628, -2.49635236, -1.32152865],
       [-1.47297026, -0.16765952,  0.04728547,  0.09766701, -0.90890988,
         1.89545411, -0.14251002,  0.71037524,  0.88750635,  0.64037105],
       [ 0.73479244,  1.21844573,  1.52764066, -0.96303431, -0.89581913,
        -0.54006801,  0.63775696,  0.75680203, -0.74836689, -0.11613013],
       [-0.59126211,  0.07754291,  0.41377188, -4.02921844, -0.65378227,
         0.09208099, -0.93864826, -0.20896442,  0.21979628,  0.99372448],
       [-1.50193608,  0.51057318,  1.17887181,  1.70768268,  0.33034431,
        -1.14658574,  1.17370376, -1.90004594

### Mean

In [33]:
G.mean()

-0.093640650928184785

### Variance

In [34]:
G.var()

1.1769978620816639

## Matrix multiplication

In [35]:
A = np.array([[1, 2, 3],[4, 5, 6]])
B = np.array([[1, 2], [3, 4], [5, 6]])
A.dot(B)

array([[22, 28],
       [49, 64]])

## Matrix elementwise multiplication

In [36]:
A = np.array([[1, 2, 3],[4, 5, 6], [7, 8, 9]])
B = np.array([[1, 2, 3],[4, 5, 6], [7, 8, 9]])
A * B

array([[ 1,  4,  9],
       [16, 25, 36],
       [49, 64, 81]])

## Matrix inverse

In [37]:
A = np.array([[1,2],[3,4]])
Ainv = np.linalg.inv(A)
Ainv

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [38]:
Ainv.dot(A)

array([[  1.00000000e+00,   0.00000000e+00],
       [  1.11022302e-16,   1.00000000e+00]])

## Matrix Determinant

In [39]:
np.linalg.det(A)

-2.0000000000000004

## Matrix diagonal

In [40]:
np.diag(A)

array([1, 4])

# Solving a linear system

## Problem:   Ax = b
## Solution:   Ainv A x = x = Ainv b

In [41]:
A = np.array([[1,2],[3,4]])
b = np.array([1, 2])
x = np.linalg.inv(A).dot(b)
x

array([ 0. ,  0.5])

In [42]:
x = np.linalg.solve(A, b)
x

array([ 0. ,  0.5])

# Example problem

### The admission fee at a small fair is \$1.50 for children and \$4.00 for adults. On a certain day, 2200 people enter the fair and \$5050 is collected. How many children and how many adults attended?

#### Let:
#### X1 = number of children
#### X2 = number of adults

#### X1 + X2 = 2200

#### 1.5X1 + 4X2 = 5050

In [43]:
A = np.array([[1, 1],[1.5, 4]])
b = np.array([2200, 5050])

np.linalg.solve(A, b)

array([ 1500.,   700.])